### ISSUE TITLE
unconstrained geometry incorrectly restored after horizontal/vertical maximize

### ISSUE THREAD

**Issue by mvduin** [awesomeWM/awesome#2062]
awesome 4.0 and 4.2 (but beware of #2061 in 4.2)

**How to reproduce the issue:**

- Horizontally or vertically maximize a window
- Move or resize the window in the other direction
- Unmaximize the window

**Actual result:**

Both horizontal and vertical size and position of the window is restored, thus undoing the manual positioning done by the user.

**Expected result:**

Only the size and position in the direction in which the window was maximized should be restored.

**Comment by psychon** [comment]
Assigning to @Elv13 since he had all the fun with this and I'm not actually sure where this restoration is implemented (was it `awful.placement`?).

**Comment by mvduin** [comment]
Just to clarify, I'm using floating layout.

**Comment by Elv13** [comment]
@blueyed What's your opinion on this? Didn't the current behavior is less or more what you wanted?

**Comment by blueyed** [comment]
@Elv13 
This is not really related to immobilized clients, is it?

This would rather prevent resizing the floating window in the first place then, which you might even expect to happen after all?

**Comment by Elv13** [comment]
> This is not really related to immobilized clients, is it?

No, I was talking about the changes you asked for v4.1 that restored some extra properties like the screen.

> This would rather prevent resizing the floating window in the first place then, which you might even expect to happen after all?

I don't know, I don't use that feature. Someone will claim it's a regression, others wont care. You just added a `immobilized_horizontal` and `immobilized_vertical`, so I guess blocking the resizing of partially maximized clients would render it useless (as 2 properties). Also, resizing partially maximized clients makes some sense "because why not?".

I will let you choose how to go ahead on this one.

**Comment by blueyed** [comment]
I think it is a missing feature of `immobilized_horizontal` and `immobilized_vertical` to apply to resizing a floating window, i.e. when `immobilized_horizontal` you should be able to only resize it vertically.

Adding this would solve this issue indirectly, because you would not be able to resize it where it is maximized.
What do you think?

**Comment by mvduin** [comment]
I'm a bit confused here. If I understand correctly you're talking about resizing not being constrained when it should have been? I had previously opened #2061 for that since I thought #2036 was specific to requests by the client, but they got merged.

**Comment by blueyed** [comment]
https://github.com/awesomeWM/awesome/commit/678ead26341bd2e0aa27a4b4707b562b1b5c770f fixes this apparently only for when clients try to resize themselves, while I would have expected that requests from the user would be blocked then, too - I am not resizing floating windows myself usually, so did not pay attention to it.

**Comment by mvduin** [comment]
I've decided to have go at digging into these problems and fixing them instead of continuing to use awesome 4.0 (which was my solution so far, since later versions were too broken for me to be usable), see pull request #2767 to finally fix #2061/#2036.

As for restore problems, I noticed the situation is far worse than I thought and not limited to maximize_horizontal/vertical. For example if you maximize a window, then set it fullscreen, then unmaximize it, it will appear to exit fullscreen mode (even though the client's `fullscreen` property is still true) and appear maximized (even though the `maximized` property is false), and the original geometry is lost.

I'm still not very familiar with awesome's codebase, and I have no idea why awful.placement was designed to work the way it does. For now I've just fixed the issue for myself by rewriting the maximization logic to use its own (really simple) restore mechanism instead of awful.placement's: https://github.com/mvduin/awesome/commit/271efc699c4a
